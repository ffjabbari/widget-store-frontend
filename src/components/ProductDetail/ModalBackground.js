import styled from 'styled-components';

// background to fade page behind open cart
const ScreenBackground = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  background: rgba(0, 0, 0, 0.5);
  z-index: 1040;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  overflow-y: auto;
  overflow-x: hidden;

  @media screen and (max-width: 991.9px) {
    justify-content: flex-start;
  }
`;

export default ScreenBackground;
