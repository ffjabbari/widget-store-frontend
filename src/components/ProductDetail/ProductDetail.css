.ProductDetail .modal-body .details {
  display: flex;
  justify-content: space-between;
  flex-direction: column;
  min-height: 100%;
}

.ProductDetail .price {
  font-weight: bold;
}

.ProductDetail input[type='number'] {
  width: 3rem;
  margin-right: 1rem;
  margin-left: 0.3rem;
}

@keyframes fadeInKey {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}

@keyframes fadeOutKey {
  0% {
    opacity: 1;
  }
  100% {
    opacity: 0;
  }
}

.fadeIn {
  animation: fadeInKey 0.8s ease-out 1;
}

.fadeOut {
  animation: fadeOutKey 0.8s ease-out 1;
}
