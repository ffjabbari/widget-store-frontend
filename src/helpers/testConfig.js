/** Test data setup/teardown helper hooks */

export const TEST_DATA = {};

export const populateTestDataHook = testData => {
  testData.product = {
    byline: 'Latest generation smart televison 2021',
    date_added: '2020-12-29T08:00:00.000Z',
    description:
      'Lorem ipsum dolor sit, amet consectetur adipisicing elit. Nihil, voluptates enim nam voluptatum, quia doloribus quis sequi, porro consequatur excepturi eos error voluptas nobis modi atque magni commodi amet placeat.',
    discount: '0.00',
    image_url:
      'https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg',
    name: 'Samsung XG-900 55 inch LCD TV',
    price: '400.40',
    product_id: 1,
    quantity: 100,
    rating: null,
  };

  testData.product2 = {
    byline: 'A good remote.',
    discount: '0.15',
    image_url:
      'https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg',
    name: 'Logitech Singularity Universal Remote',
    price: '40.20',
    product_id: 2,
    rating: null,
    quantity: 1,
  };

  testData.userProfileData = {
    data: {
      user: {
        user_id: 1,
        username: 'widgetLord',
        email: 'foo@gmail.com',
        first_name: 'bob',
        last_name: 'bobert',
        address: '123 Hill Ave.',
        address_line2: null,
        city: 'Big City',
        state: 'Texas',
        postal_code: '33333',
        phone_number: null,
        avatar_url: null,
        orders: [
          {
            order_id: 1,
            order_date: '2021-01-24T08:00:00.000Z',
            status: 'Pending',
          },
        ],
      },
    },
  };

  testData.CustomerInfoFormsValues = {
    first_name: 'Test',
    last_name: 'User',
    email: 'foo@gmail.com',
    address: '123 Main St',
    address_line2: '',
    city: 'Big City',
    state: 'TX',
    postal_code: '39488',
    phone_number: '(555) 555-5555',
    set_default: false,
  };

  testData.cart = {
    items: {
      2: {
        byline: 'The only remote you will need',
        date_added: '2021-01-24T08:00:00.000Z',
        description: 'Lorem ipsum dolor sit, placeat.',
        discount: '0.15',
        image_url:
          'https://cnet3.cbsistatic.com/img/SB-Lc9lni02k-GQeFZlVGII6cpg=/1070x602/2010/08/03/57369690-bb77-11e2-8a8e-0291187978f3/33981226-2-1333-OVR-1.jpg',
        name: 'Logitech Singularity Universal Remote',
        price: '40.20',
        product_id: 2,
        quantity: 2,
        rating: null,
      },
    },
    numCartItems: 1,
    subtotal: '34.17',
  };
};
