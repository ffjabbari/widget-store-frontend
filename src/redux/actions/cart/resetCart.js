/** Reset action creator. */

import { RESET_CART } from '../types';

function resetCart() {
  return { type: RESET_CART };
}

export default resetCart;
