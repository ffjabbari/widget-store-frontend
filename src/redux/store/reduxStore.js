/*
 * Uses redux-persist to persist state.
 * Uses redux-thunk to allow async operations.
 */

import { composeWithDevTools } from 'redux-devtools-extension';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import autoMergeLevel2 from 'redux-persist/lib/stateReconciler/autoMergeLevel2';
import thunk from 'redux-thunk';
import { createStore, applyMiddleware } from 'redux';

// local imports
import root from '../reducers/root';

const persistConfig = {
  key: 'root',
  storage,
  // allow any new state properties to persist state rehydration
  stateReconciler: autoMergeLevel2,
};

const persistedReducer = persistReducer(persistConfig, root);

export const store = createStore(
  persistedReducer,
  composeWithDevTools(applyMiddleware(thunk))
);

export const persistedStore = persistStore(store);

export const testStore = createStore(root);
