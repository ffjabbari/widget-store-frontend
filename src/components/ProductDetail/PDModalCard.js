import { useState } from 'react';
import { useLocation, useHistory } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import {
  Button,
  ModalHeader,
  ModalBody,
  ModalFooter,
  UncontrolledTooltip,
} from 'reactstrap';
import { motion } from 'framer-motion';

// local imports
import ModalCard from './ModalCard';
import Pricing1 from '../common/Pricing/Pricing1';
import CartIconBadge from '../common/CartIconBadge/CartIconBadge';
import addProduct from '../../redux/actions/cart/addProduct';
import { getPathRoot, animateVariant } from '../../helpers/helpers';

const PDModal = ({ handleClose, product, innerContentOnly, disabled }) => {
  const location = useLocation();
  const pathRoot = getPathRoot(location.pathname);
  const history = useHistory();
  const dispatch = useDispatch();

  // state
  const numCartItems = useSelector(state => state.cart.numCartItems);
  const [animClass, setBtnAnimation] = useState('default');

  // add to cart form
  const INIT_FORM_STATE = { quantity: disabled ? product.quantity : 1 };
  const [formData, setFormData] = useState(INIT_FORM_STATE);

  const handleChange = e => {
    const { name, value } = e.target;
    setFormData(formData => ({
      ...formData,
      [name]: +value,
    }));
  };

  const handleAddToCart = e => {
    e.preventDefault();
    dispatch(addProduct({ ...product, quantity: formData.quantity }));
    animateVariant(setBtnAnimation, 1000);
  };

  const goToCart = () => {
    history.push(`${pathRoot}/cart`);
  };

  // modal close btn
  const closeBtn = <button className="close-modal close">&times;</button>;

  const modalInner = (
    <>
      <ModalHeader toggle={handleClose} close={closeBtn}>
        <div>{product.name}</div>
      </ModalHeader>
      <ModalBody className="p-0 row">
        <div className="col-12 col-lg-6">
          <img
            className={`ProductDetail-img img-fluid`}
            src={product.image_url}
            alt={product.name}
          />
        </div>

        <div className="col-12 col-lg-6">
          <div className="details p-2">
            <h5 className="mb-4 text-center pt-4">{product.byline}</h5>
            <p>{product.description}</p>
            <div className="text-right pb-3 pr-5 mr-5">
              <Pricing1 product={product} />
            </div>
          </div>
        </div>
      </ModalBody>
      <ModalFooter>
        <form onSubmit={handleAddToCart}>
          <label htmlFor="quantity">Quantity</label>
          <input
            type="number"
            id="quantity"
            name="quantity"
            min="1"
            placeholder="1"
            onChange={handleChange}
            value={formData.quantity}
            disabled={disabled}
          />
          <Button color="primary" disabled={disabled}>
            Add To Cart
          </Button>
        </form>
        <motion.div
          variants={{
            active: {
              scale: 1.6,
              rotateZ: 360,
              '--btnAnimColor': 'goldenrod',
              transition: 0.3,
            },
            default: {},
          }}
          animate={animClass}
          style={{ '--btnAnimColor': 'white' }}
        >
          <Button
            color="secondary"
            className={animClass}
            style={{ color: 'var(--btnAnimColor)' }}
            id="cart-button"
            onClick={goToCart}
            disabled={disabled || pathRoot === '/checkout'}
          >
            <CartIconBadge
              numCartItems={numCartItems}
              size="lg"
              className={`large`}
            />
          </Button>
          <UncontrolledTooltip placement="top" target="cart-button">
            Go To Shopping Cart
          </UncontrolledTooltip>
        </motion.div>
        <Button color="secondary" className="close-modal">
          Cancel
        </Button>
      </ModalFooter>
    </>
  );

  // for placing content inside bootstrap modal
  if (innerContentOnly) return modalInner;

  return (
    <ModalCard className="ProductDetail" color="var(--dark)">
      {modalInner}
    </ModalCard>
  );
};

export default PDModal;
