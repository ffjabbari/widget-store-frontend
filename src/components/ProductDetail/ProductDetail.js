import { useParams, useHistory, Link, useLocation } from 'react-router-dom';
import React, { useState } from 'react';
import {
  Button,
  Modal,
  ModalHeader,
  ModalBody,
  ModalFooter,
  UncontrolledTooltip,
} from 'reactstrap';
import { useDispatch, useSelector } from 'react-redux';
import { motion } from 'framer-motion';

// local imports
import APIRequest from '../../hooks/apiHook';
import Pricing1 from '../common/Pricing/Pricing1';
import CartIconBadge from '../common/CartIconBadge/CartIconBadge';
import addProduct from '../../redux/actions/cart/addProduct';
import { getPathRoot, animateVariant } from '../../helpers/helpers';
import './ProductDetail.css';

const ProductDetail = () => {
  const { id } = useParams();
  const history = useHistory();
  const dispatch = useDispatch();
  const [animClass, setBtnAnimation] = useState('default');
  const numCartItems = useSelector(state => state.cart.numCartItems);
  const location = useLocation();
  const pathRoot = getPathRoot(location.pathname);

  // add to cart form
  const INIT_FORM_STATE = { quantity: 1 };
  const [formData, setFormData] = useState(INIT_FORM_STATE);

  // close modal function
  const handleClose = e => {
    return history.push(pathRoot);
  };

  // modal variables
  const [modal] = useState(true);
  const closeBtn = (
    <button className="close" onClick={handleClose}>
      &times;
    </button>
  );

  // load product data
  const { loading, error, response } = APIRequest(`/products/${id}`, 'get');

  // if no product found go to path root i.e. /shop, /new, /deals
  if (error) {
    history.replace(pathRoot);
    return null;
  }
  if (loading) return null;

  const product = response.product;

  const handleChange = e => {
    const { name, value } = e.target;
    setFormData(formData => ({
      ...formData,
      [name]: +value,
    }));
  };

  const handleAddToCart = e => {
    e.preventDefault();
    const copy = { ...product };
    delete copy.byline;
    delete copy.description;
    delete copy.date_added;
    delete copy.rating;
    dispatch(addProduct({ ...copy, quantity: formData.quantity }));
    animateVariant(setBtnAnimation, 1000);
  };

  return (
    <Modal
      isOpen={modal}
      toggle={handleClose}
      className="ProductDetail"
      size="xl"
      centered
    >
      <ModalHeader toggle={handleClose} close={closeBtn}>
        <div>{product.name}</div>
      </ModalHeader>
      <ModalBody className="p-0 row">
        <div className="col-12 col-lg-6">
          <img
            className={`ProductDetail-img img-fluid`}
            src={product.image_url}
            alt={product.name}
          />
        </div>

        <div className="col-12 col-lg-6">
          <div className="details p-2">
            <h5 className="mb-4 text-center pt-4">{product.byline}</h5>
            <p>{product.description}</p>
            <div className="text-right pb-3 pr-5 mr-5">
              <Pricing1 product={product} />
            </div>
          </div>
        </div>
      </ModalBody>
      <ModalFooter>
        <form onSubmit={handleAddToCart}>
          <label htmlFor="quantity">Quantity</label>
          <input
            type="number"
            id="quantity"
            name="quantity"
            min="1"
            placeholder="1"
            onChange={handleChange}
            value={formData.quantity}
          />
          <Button color="primary">Add To Cart</Button>
        </form>
        <Link to={`${pathRoot}/cart`}>
          <motion.div
            variants={{
              active: {
                scale: 1.6,
                rotateZ: 360,
                '--btnAnimColor': 'goldenrod',
                transition: 0.3,
              },
              default: {},
            }}
            animate={animClass}
            style={{ '--btnAnimColor': 'white' }}
          >
            <Button
              color="secondary"
              className={animClass}
              style={{ color: 'var(--btnAnimColor)' }}
              id="cart-button"
            >
              <CartIconBadge
                numCartItems={numCartItems}
                size="lg"
                className={`large`}
              />
            </Button>
            <UncontrolledTooltip placement="top" target="cart-button">
              Go To Shopping Cart
            </UncontrolledTooltip>
          </motion.div>
        </Link>
        <Button color="secondary" onClick={handleClose}>
          Cancel
        </Button>
      </ModalFooter>
    </Modal>
  );
};

export default ProductDetail;
