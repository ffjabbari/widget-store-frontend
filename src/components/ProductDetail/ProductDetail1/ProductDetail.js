/** Product Detail Modal Type 1
 *
 * Shows product details on a Bootstrap (reactstrap) modal.
 */

import { useParams, useHistory, useLocation } from 'react-router-dom';
import React, { useState } from 'react';
import { Modal } from 'reactstrap';

// local imports
import APIRequest from '../../../hooks/apiHook';
import { getPathRoot } from '../../../utils/helpers';
import PDModalCard from '../modal/PDModalCard';
import '../ProductDetail.css';

const ProductDetail = () => {
  const { id } = useParams();
  const history = useHistory();
  const location = useLocation();
  const pathRoot = getPathRoot(location.pathname);

  // close modal function.
  // if filterTarget is true only elements with a
  // class of "close-modal" will close the modal.
  const handleClose = (e, filterTarget) => {
    if (filterTarget && !e.target.classList.contains('close-modal')) return;
    return history.push(pathRoot);
  };

  // modal variables
  const [modal] = useState(true);

  // load product data
  const { loading, error, response } = APIRequest(`/products/${id}`, 'get');

  // if no product found go to path root i.e. /shop, /new, /deals
  if (error) {
    history.replace(pathRoot);
    return null;
  }
  if (loading) return null;

  const product = response.product;

  return (
    <Modal
      isOpen={modal}
      toggle={handleClose}
      className="ProductDetail"
      size="xl"
      centered
    >
      <div onClick={e => handleClose(e, true)}>
        <PDModalCard
          product={product}
          handleClose={handleClose}
          innerContentOnly="true"
        />
      </div>
    </Modal>
  );
};

export default ProductDetail;
