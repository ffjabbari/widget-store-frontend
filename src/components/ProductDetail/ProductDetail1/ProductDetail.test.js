/** ProductDetail tests */
import { act, fireEvent } from '@testing-library/react';
import axios from 'axios';

// local imports
import { renderWithStore } from '../../../utils/testHelpers';
import { testStore } from '../../../redux/store/reduxStore';
import ProductDetail from './ProductDetail';
import resetCart from '../../../redux/actions/cart/resetCart';
import { TEST_DATA, populateTestDataHook } from '../../../utils/testConfig';

jest.mock('axios');

const mockUseHistoryReplace = jest.fn(),
  mockUseHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    replace: mockUseHistoryReplace,
    push: mockUseHistoryPush,
  }),
}));

beforeAll(() => {
  populateTestDataHook(TEST_DATA);
});

beforeEach(() => {
  axios.mockResolvedValue({
    data: {
      product: TEST_DATA.product,
    },
  });
});

afterEach(() => {
  jest.resetAllMocks();
  testStore.dispatch(resetCart());
});

test('renders ProductDetail', async () => {
  await act(async () => {
    renderWithStore(<ProductDetail />);
  });
});

test('ProductDetail snapshot', async () => {
  let asFragment;
  await act(async () => {
    const resp = renderWithStore(<ProductDetail />);
    asFragment = resp.asFragment;
  });
  expect(asFragment()).toMatchSnapshot();
});

test('Quantity input in document', async () => {
  let getByLabelText;
  await act(async () => {
    const resp = renderWithStore(<ProductDetail />);
    getByLabelText = resp.getByLabelText;
  });
  // test quantity input present with value of '1'
  const quantityInput = getByLabelText('Quantity');
  expect(quantityInput).toBeInTheDocument();
  expect(quantityInput.value).toBe('1');
});

test('Add to cart adds to store.cart data', async () => {
  let getByText;
  await act(async () => {
    const resp = renderWithStore(<ProductDetail />);
    getByText = resp.getByText;
  });
  // click add to cart button
  const addBtn = getByText('Add To Cart');
  act(() => {
    fireEvent.click(addBtn);
  });
  // test item data in store.cart
  const cart = testStore.getState().cart;
  expect(cart.items).toHaveProperty('1');
  expect(cart.items[1].quantity).toBe(1);
  expect(cart.items[1].price).toBe('400.40');
});

test('Add to cart can add multiple quantity to store.cart data', async () => {
  let getByText, getByLabelText;
  await act(async () => {
    const resp = renderWithStore(<ProductDetail />);
    getByText = resp.getByText;
    getByLabelText = resp.getByLabelText;
  });
  // set item quantity to add to cart to "3"
  const quantityInput = getByLabelText('Quantity');
  fireEvent.change(quantityInput, { target: { value: '3' } });
  // add to cart
  const addBtn = getByText('Add To Cart');
  act(() => {
    fireEvent.click(addBtn);
  });
  // check store.cart item quantity
  const cart = testStore.getState().cart;
  expect(cart.items).toHaveProperty('1');
  expect(cart.items[1].quantity).toEqual(3);
});

test('history replaces url with path root with get product error', () => {
  axios.mockImplementation(() => {
    throw new Error();
  });

  renderWithStore(<ProductDetail />);

  expect(mockUseHistoryReplace.mock.calls.length).toBe(1);
  expect(mockUseHistoryReplace.mock.calls[0][0]).toBe('/');
});

test('history pushes path root with modal close', async () => {
  let getByRole;
  await act(async () => {
    const resp = renderWithStore(<ProductDetail />);
    getByRole = resp.getByRole;
  });

  const cancelButton = getByRole('button', { name: /cancel/i });
  fireEvent.click(cancelButton);

  expect(mockUseHistoryPush.mock.calls.length).toBe(1);
  expect(mockUseHistoryPush.mock.calls[0][0]).toBe('/');
});
