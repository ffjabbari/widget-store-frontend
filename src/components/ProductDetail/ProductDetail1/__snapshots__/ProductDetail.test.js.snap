// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetail snapshot 1`] = `<DocumentFragment />`;
